### ISSUE TITLE
MTKClient crashed when trying to unlock bootloader.

### ISSUE THREAD

**Issue by ArkYuki** [bkerler/mtkclient#1167]
I'm trying to unlock my Redmi 6A. Because I can't access my windows PC, I use Debian instead.
python threw me this error. I have tried install libusb-1.0-0-dev, but still.
```
$ python3 mtk.py da seccfg unlock
MTK Flash/Exploit Client Public V2.0.1 (c) B.Kerler 2018-2024

DAXFlash - HW-CODE         : 0x766
DAXFlash - HWSUB-CODE      : 0x8A00
DAXFlash - HW-VERSION      : 0xCA00
DAXFlash - SW-VERSION      : 0x0
DAXFlash - CHIP-EVOLUTION  : 0x0
DAXFlash - DA-VERSION      : 1.0
XFlashExt - Detected V4 Lockstate
Sej - HACC init
Sej - HACC run
Sej - HACC terminate
Sej - HACC init
Sej - HACC run
Sej - HACC terminate
Sej - HACC init
Sej - HACC run
Sej - HACC terminate
Sej - HACC init
Sej - HACC run
Sej - HACC terminate
Traceback (most recent call last):
  File "/home/arkyuki/mtkclient/mtk.py", line 1016, in <module>
    main()
  File "/home/arkyuki/mtkclient/mtk.py", line 1012, in main
    mtk = Main(args).run(parser)
          ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/arkyuki/mtkclient/mtkclient/Library/mtk_main.py", line 664, in run
    da_handler.handle_da_cmds(mtk, cmd, self.args)
  File "/home/arkyuki/mtkclient/mtkclient/Library/DA/mtk_da_handler.py", line 877, in handle_da_cmds
    v = mtk.daloader.seccfg(args.flag)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/arkyuki/mtkclient/mtkclient/Library/DA/mtk_daloader.py", line 392, in seccfg
    return self.xft.seccfg(lockflag)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/arkyuki/mtkclient/mtkclient/Library/DA/xflash/extension/xflash.py", line 654, in seccfg
    if self.xflash.writeflash(addr=partition.sector * self.mtk.daloader.daconfig.pagesize,
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/arkyuki/mtkclient/mtkclient/Library/DA/xflash/xflash_lib.py", line 938, in writeflash
    if os.path.exists(filename):
       ^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen genericpath>", line 19, in exists
TypeError: stat: path should be string, bytes, os.PathLike or integer, not NoneType

```

**Comment by ThaYke** [comment]
Same here after trying the commit which was meant to fix the "jumping to extensions"-bug.

**Comment by weskerty** [comment]
try cloning the repository and not using download?

**Comment by ThaYke** [comment]
Actually, I used git pull, and reinstalled, as I had it cloned on my machine before, already. Do you think that could lead to problems?

**Comment by ThaYke** [comment]
Just tried it on another machine (Fedora 39) with a freshly cloned repo. Samoe error.

**Comment by ArkYuki** [comment]
Maybe it's a permission issue? Try running python with root?
I can't get my computer now.

**Comment by ArkYuki** [comment]
> try cloning the repository and not using download?

I exactly used git clone.

**Comment by rickitaly** [comment]
same problem with a Xiaomi Smart Clock x04g
tried from bot Raspian OS and Windows, same result!

**Comment by undervers** [comment]
Hi. I am unbricking my friend pnone and i had this issue too. I dont know code of this project but it seems to me it is a simple bug.

It is all about this:

/DA/xflash/xflash_lib.py", line 938, in writeflash

def writeflash(self, addr, length, filename: str = "", offset=0, parttype=None, wdata=None, display=True):
        self.mtk.daloader.progress.clear()
        fh = None
        fill = 0
        if filename != "":
            print(filename)
            if os.path.exists(filename):
                fsize = os.stat(filename).st_size
                length = min(fsize, length)
                if length % 512 != 0:
                    fill = 512 - (length % 512)
                    length += fill
                fh = open(filename, "rb")
                fh.seek(offset)
            else:
                self.error(f"Filename doesn't exists: {filename}, aborting flash write.")
                return False


As we can see writeFlash function have filename parameter which is empty string by default. This fucnction called here:

DA/xflash/extension/xflash.py", line 654, in seccfg

if self.xflash.writeflash(addr=partition.sector * self.mtk.daloader.daconfig.pagesize,
                                  length=len(writedata),
                                  filename=None, wdata=writedata, parttype="user", display=True):

As you can see parameter "filename" set in "None" not in "". So for workaround you need just replace "None" with "".

**Comment by ThaYke** [comment]
> Hi. I am unbricking my friend pnone and i had this issue too. I dont know code of this project but it looks like a simple bug.
> 
> It is all about this:
> 
> /DA/xflash/xflash_lib.py", line 938, in writeflash
> 
> ```
> def writeflash(self, addr, length, filename: str = "", offset=0, parttype=None, wdata=None, display=True):
>         self.mtk.daloader.progress.clear()
>         fh = None
>         fill = 0
>         if filename != "":
>             if os.path.exists(filename):
>                 fsize = os.stat(filename).st_size
>                 length = min(fsize, length)
>                 if length % 512 != 0:
>                     fill = 512 - (length % 512)
>                     length += fill
>                 fh = open(filename, "rb")
>                 fh.seek(offset)
>             else:
>                 self.error(f"Filename doesn't exists: {filename}, aborting flash write.")
>                 return False
> ```
> 
> As we can see writeFlash function have filename parameter which is empty string by default. This function called here:
> 
> DA/xflash/extension/xflash.py", line 654, in seccfg
> 
> ```
> if self.xflash.writeflash(addr=partition.sector * self.mtk.daloader.daconfig.pagesize,
>                                   length=len(writedata),
>                                   filename=None, wdata=writedata, parttype="user", display=True):
> ```
> 
> As you can see parameter "filename" set in "None" not in "". So for workaround just replace 'filename=None' with 'filename=""'.

Thank you for the hint, this worked for me.

**Comment by rickitaly** [comment]
I solved by using old 1.* version, but indeed solution suggested should be working!

**Comment by bkerler** [comment]
fixed with [8fff2ab](https://github.com/bkerler/mtkclient/commit/8fff2aba850e280e2f22da938b67cbf7da3d2b5c)

**Comment by hacker-hackman** [comment]
@ArkYuki 
So, did you successfully unlock the bootloader on your Redmi 6A?
